src/index.js

```diff
@@ -1,83 +1,135 @@
 'use strict'
 
 const net = require('net')
-const toPull = require('stream-to-pull-stream')
 const mafmt = require('mafmt')
 const withIs = require('class-is')
-const includes = require('lodash.includes')
-const isFunction = require('lodash.isfunction')
-const Connection = require('interface-connection').Connection
-const once = require('once')
-const debug = require('debug')
-const log = debug('libp2p:tcp:dial')
-
+const errCode = require('err-code')
+const log = require('debug')('libp2p:tcp')
+const toConnection = require('./socket-to-conn')
 const createListener = require('./listener')
+const { AbortError } = require('abortable-iterator')
+const { CODE_CIRCUIT, CODE_P2P } = require('./constants')
 
-function noop () {}
-
+/**
+ * @class TCP
+ */
 class TCP {
-  dial (ma, options, callback) {
-    if (isFunction(options)) {
-      callback = options
-      options = {}
-    }
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {Upgrader} options.upgrader
+   */
+  constructor ({ upgrader }) {
+    this._upgrader = upgrader
```

Comment:
should we `assert` that we receive the `upgrader`

Reply:
Added the assert and a test. I included a link to the Upgrader definition in the interface-transport repo for UX.